realm/src/main/java/io/realm/Realm.java

```diff
@@ -1872,4 +1890,24 @@ public static Object getDefaultModule() {
         void execute(Realm realm);
     }
 
+    /**
+     * Encapsulates an async {@link RealmQuery}.
+     * <p>
+     * This will run the {@link RealmQuery} on a worker thread, then invoke this callback on the caller thread
+     */
+    public interface QueryCallback<T extends RealmResults<? extends RealmObject>> {
```

Comment:
Couldn't this be `QueryCallback<E extends RealmObject>` with `onSuccess(RealmResult<E> results)` ? 
Or are we expecting to extend `RealmResults`?